### PR TITLE
Improve Period range indexing

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java [hunk 1]
```diff
@@ -50,6 +54,8 @@
 		"http://hl7.org/fhir/StructureDefinition/*")));
 
 	public static final String DEFAULT_WEBSOCKET_CONTEXT_PATH = "/websocket";
+	protected static final String DEFAULT_PERIOD_INDEX_START_OF_TIME = "1700-01-01";
```
Comment (fil512) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java:57]: Why not 0000-01-01 and 9999-12-31 ?

Comment (fil512) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java:57]: Someone might want to store genealogy data (or even ancestral disease information) that predates this date.

Comment (fil512) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java:57]: On the other hand, we should check that these dates work okay on the main databases, esp the timestamp versions of them.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 2]
```diff
@@ -677,6 +677,16 @@ private void addDate_Period(String theResourceType, Set<ResourceIndexedSearchPar
 		String endAsString = extractValueAsString(myPeriodEndValueChild, theValue);
 
 		if (start != null || end != null) {
+
+			if (start == null) {
+				start = myModelConfig.getPeriodIndexStartOfTime().getValue();
+				startAsString = myModelConfig.getPeriodIndexStartOfTime().getValueAsString();
+			}
+			if (end == null) {
+				end = myModelConfig.getPeriodIndexEndOfTime().getValue();
```
Comment (fil512) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:686]: I'm surprised you didn't need the DatePredicate logic.  I guess those queries just never worked?

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java [hunk 3]
```diff
@@ -1118,143 +1118,6 @@ public void testCreateWrongType() {
...
-		params.add(Encounter.SP_DATE, new DateRangeParam(null, "2001-01-03"));
-		params.add(Encounter.SP_IDENTIFIER, new TokenParam("testDatePeriodParam", "02"));
-		encs = toList(myEncounterDao.search(params));
-		assertEquals(1, encs.size());
-
-		params = new SearchParameterMap();
-		params.add(Encounter.SP_DATE, new DateRangeParam("2001-01-01", "2001-01-03"));
-		params.add(Encounter.SP_IDENTIFIER, new TokenParam("testDatePeriodParam", "02"));
-		// encs = toList(ourEncounterDao.search(params));
-		// assertEquals(1, encs.size());
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java:1142]: Did this case just not work, so the assert was commented out?

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 4]
```diff
@@ -1690,6 +1691,338 @@ public void testSearchDateWrongParam() {
...
+		params.add(Encounter.SP_DATE, new DateRangeParam(null, "2001-01-03"));
+		params.add(Encounter.SP_IDENTIFIER, new TokenParam("testDatePeriodParam", "02"));
+		encs = toList(myEncounterDao.search(params));
+		assertEquals(1, encs.size());
+
+		params = new SearchParameterMap();
+		params.add(Encounter.SP_DATE, new DateRangeParam("2001-01-01", "2001-01-03"));
+		params.add(Encounter.SP_IDENTIFIER, new TokenParam("testDatePeriodParam", "02"));
+		// encs = toList(ourEncounterDao.search(params));
+		// assertEquals(1, encs.size());
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java:1908]: Does this work now if you uncomment it?  Should it work?

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ModelConfig.java:57]: Did a bit of digging into supported limits in databases:

<p>
<ul>
<li>JDBC Timestamp Datatype Low Value -4713 and High Value 9999</li>
<li>MySQL 8: the range for DATETIME values is '1000-01-01 00:00:00.000000' to '9999-12-31 23:59:59.999999`</li>
<li>Postgresql 12: Timestamp [without time zone] Low Value 4713 BC and High Value 294276 AD</li>
<li>Oracle: Timestamp Low Value 4712 BC and High Value 9999 CE</li>
<li>H2: datetime2 Low Value -4713 and High Value 9999</li>
</ul>
</p>

I'm gonna go with 1001 - 9000 for safety

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java:1908]: Ah good call! It totally does.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java:1142]: These tests weren't actually removed, I just moved them from `FhirResourceDaoR4Test` to `FhirResourceDaoR4SearchNoFtTest.java` in order to make them easier to find (that's where all the tests for the other search datatypes were, no idea why the date ones were here!)

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:686]: Yup. I'm happy about that aspect of this change, no modifications to the search predicate logic were required.

It's not as bad as it maybe seemed- Date range searches almost all worked, there are only 2 cases that didn't: 

* Indexed period with a lower bound and no upper bound, and then a search by date where the selector is `gt ( a value > the lower bound)`
* Indexed period with an upper bound and no lower bound, and then a search by date where the selector is `lt ( a value < the upper bound)`

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4Test.java:1142]: Uncommented in the other class, and it does work now!

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java:1908]: Indeed it does!

**Pull request merged**